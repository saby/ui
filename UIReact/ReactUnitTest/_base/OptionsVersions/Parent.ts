import { Record } from 'Types/entity';
import { Control, TemplateFunction } from 'UI/Base';
import { dafault as Child1 } from './Child1';
import * as template from 'wml!ReactUnitTest/_base/OptionsVersions/Parent';

export default class Parent extends Control {
    protected _template: TemplateFunction = template;

    protected _record: Record = new Record({
        rawData: {
            field: 'parent',
        },
    });

    getChild1(): Child1 {
        return this._children.child1;
    }
}
