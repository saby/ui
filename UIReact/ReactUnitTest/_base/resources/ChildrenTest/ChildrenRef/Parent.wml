<ws:template name="child2">
   <ReactUnitTest._base.resources.ChildrenTest.ChildrenRef.Child2 name="MyControl"/>
</ws:template>

<ws:template name="child1">
   <ReactUnitTest._base.resources.ChildrenTest.ChildrenRef.Child1 scope="{{ myOptions }}">
      <ws:partial template="child2"/>
   </ReactUnitTest._base.resources.ChildrenTest.ChildrenRef.Child1>
</ws:template>

<ws:partial template="child1" myOptions="{{ _options }}"/>