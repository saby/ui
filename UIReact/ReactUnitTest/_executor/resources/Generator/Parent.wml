<div>
   <ws:for data="index, item in _items">
        <ws:partial template="{{item.itemTemplate}}" />
   </ws:for>
</div>