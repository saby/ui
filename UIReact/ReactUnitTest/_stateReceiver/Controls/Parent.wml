<div class="ReactUnitTest-stateReceiver__parent">
  <ReactUnitTest._stateReceiver.Controls.UselessChildren/>
  <ReactUnitTest._stateReceiver.Controls.Children name="children"/>
</div>
