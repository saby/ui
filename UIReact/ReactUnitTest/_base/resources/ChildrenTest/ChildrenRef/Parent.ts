import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_base/resources/ChildrenTest/ChildrenRef/Parent';

export default class Parent extends Control {
    _template: TemplateFunction = template;
}
