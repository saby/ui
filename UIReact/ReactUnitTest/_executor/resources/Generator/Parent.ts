import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_executor/resources/Generator/Parent';

export default class Parent extends Control {
    _template: TemplateFunction = template;
    protected _testOption: string;
    // условный шаблон, получить такое возможно в случае вставки в Async c динамическим templateName
    protected _items = [{ itemTemplate: [''] }, { itemTemplate: [''] }];
}
