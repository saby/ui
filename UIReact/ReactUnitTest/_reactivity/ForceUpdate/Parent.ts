import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_reactivity/ForceUpdate/Parent';

export default class Parent extends Control {
    protected _template: TemplateFunction = template;
    protected _reactiveOption: number = 0;

    changeReactiveOption(): void {
        this._reactiveOption = this._reactiveOption + 1;
    }
}
