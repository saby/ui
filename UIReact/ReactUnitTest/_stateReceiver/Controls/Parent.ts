import { Control, TemplateFunction, IControlOptions } from 'UI/Base';
// @ts-ignore
import * as template from 'wml!ReactUnitTest/_stateReceiver/Controls/Parent';

interface ITestChildren extends Control {
    getReceivedStateKey(): string;
}

export default class Parent extends Control<IControlOptions> {
    _template: TemplateFunction = template;
    _options: IControlOptions;
    _children: {
        children: ITestChildren;
    };

    getChildrenKey(): string {
        return this._children?.children.getReceivedStateKey();
    }
}
