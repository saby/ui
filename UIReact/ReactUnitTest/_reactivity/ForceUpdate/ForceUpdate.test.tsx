/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import Parent from './Parent';

describe('вызов _forceUpdate и изменение реактивного свойства', () => {
    let container;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        jest.useFakeTimers();
        // Чтобы не зависеть от реализации requestAnimationFrame для jsdom.
        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(
            setTimeout
        );
    });

    afterEach(() => {
        jest.useRealTimers();
        unmountComponentAtNode(container);
        container.remove();
        container = null;
        jest.restoreAllMocks();
    });

    test('_forceUpdate', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(1);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('реактивность', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(1);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('_forceUpdate | реактивность', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(2);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('реактивность | _forceUpdate', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(2);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('_forceUpdate | _forceUpdate | _forceUpdate', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(3);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('_forceUpdate | _forceUpdate | реактивность', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(3);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('_forceUpdate | реактивность | _forceUpdate', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(3);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('_forceUpdate | реактивность | реактивность', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(3);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('реактивность | _forceUpdate | _forceUpdate', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(3);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('реактивность | _forceUpdate | реактивность', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(3);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('реактивность | реактивность | _forceUpdate', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent._forceUpdate();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(3);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });

    test('реактивность | реактивность | реактивность', async () => {
        const beforeUpdateSpy = jest
            .spyOn<any, string>(Parent.prototype, '_beforeUpdate')
            .mockName('beforeUpdateSpy');

        let parent: Parent;
        act(() => {
            render(
                <Parent
                    ref={(v) => {
                        parent = v;
                    }}
                />,
                container
            );
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        act(() => {
            parent.changeReactiveOption();
        });

        await act(async () => {
            jest.advanceTimersByTime(0);
            jest.runOnlyPendingTimers();
        });

        expect(beforeUpdateSpy).toBeCalledTimes(3);
        expect(container).toMatchSnapshot();

        beforeUpdateSpy.mockRestore();
    });
});
