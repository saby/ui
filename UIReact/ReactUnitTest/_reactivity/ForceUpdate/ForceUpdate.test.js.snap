// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`вызов _forceUpdate и изменение реактивного свойства _forceUpdate | _forceUpdate | _forceUpdate 1`] = `
<div>
  <div
    tabindex="0"
  >
    0
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства _forceUpdate | _forceUpdate | реактивность 1`] = `
<div>
  <div
    tabindex="0"
  >
    1
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства _forceUpdate | реактивность | _forceUpdate 1`] = `
<div>
  <div
    tabindex="0"
  >
    1
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства _forceUpdate | реактивность | реактивность 1`] = `
<div>
  <div
    tabindex="0"
  >
    2
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства _forceUpdate | реактивность 1`] = `
<div>
  <div
    tabindex="0"
  >
    1
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства _forceUpdate 1`] = `
<div>
  <div
    tabindex="0"
  >
    0
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства реактивность | _forceUpdate | _forceUpdate 1`] = `
<div>
  <div
    tabindex="0"
  >
    1
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства реактивность | _forceUpdate | реактивность 1`] = `
<div>
  <div
    tabindex="0"
  >
    2
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства реактивность | _forceUpdate 1`] = `
<div>
  <div
    tabindex="0"
  >
    1
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства реактивность | реактивность | _forceUpdate 1`] = `
<div>
  <div
    tabindex="0"
  >
    2
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства реактивность | реактивность | реактивность 1`] = `
<div>
  <div
    tabindex="0"
  >
    3
  </div>
</div>
`;

exports[`вызов _forceUpdate и изменение реактивного свойства реактивность 1`] = `
<div>
  <div
    tabindex="0"
  >
    1
  </div>
</div>
`;
